openstack/identity/identity.go

```diff
@@ -283,16 +277,24 @@ type ValidAdmin struct {
 // This custom handler allows us to more cleanly return an error and response,
 // and pass some package level context into the handler.
 type Handler struct {
-	Client        *gophercloud.ServiceClient
-	Next          http.Handler
-	ValidServices []ValidService
-	ValidAdmins   []ValidAdmin
+	Client         *gophercloud.ServiceClient
+	Next           http.Handler
+	ValidServices  []ValidService
+	ValidAdmins    []ValidAdmin
+	tokenGetResult getResult
 }
 
 // ServeHTTP satisfies the http handler interface.
 // It will check to make sure that the api caller is validated with
 // keystone before allowing the next handler in the chain to be called.
 func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+
+	token := r.Header["X-Auth-Token"]
+	if token != nil {
+		res := v3tokens.Get(h.Client, token[0])
+		h.tokenGetResult = getResult{res}
```

Comment:
To be honest I don't really understand how all this identity stuff works but I'm concerned about what would happen if we have multiple clients.  Do they send different X-Auth-Token values in their request?  If so, and we have one handler serving these multiple clients, aren't we creating a security problem here, different clients could use each other's tokens.  This would be the case even if we fixed the race above using mutexes.  If you want to cache the results of v3tokens.Get you'll need to do it in a local variable and pass it to other functions via a parameter.